### PR TITLE
[JSC] Make each IC as one big handler

### PR DESCRIPTION
**Pull request by Constellation**
#### 7d177e2ec1a6260e4272a03a4d2baf33dfd065ac
<pre>
[JSC] Make each IC as one big handler
<a href="https://bugs.webkit.org/show_bug.cgi?id=263414">https://bugs.webkit.org/show_bug.cgi?id=263414</a>
rdar://117227541

Reviewed by Keith Miller.

This is a part of handler IC transition. In this patch, we make all IC as one big handler IC in each place.
So, each IC&apos;s launching code becomes the final form, which loads a handler and invoke it.
And now, each IC only has one handler which includes everything. The approach in handler IC will split this handler into chain of handlers,
and sharing these handlers across the IC sites.

Now, we need to have additional one level load for each IC, and we would like to alleviate the effect of this. To make it work well,
in this patch, we introduce ButterflyArray, and remove one level of load to get StructureStubInfo. Now StructureStubInfo is directly embedded
into JITData, and we can compute the pointer without load from that. This removes one load, and instead we introduce one load for handler.
So, while we are not getting the final handler IC benefit, we can keep perf neutral while we are continuously doing incremental work towards
handler IC.

* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::SharedJITStubSet::getMegamorphic): Deleted.
(JSC::SharedJITStubSet::setMegamorphic): Deleted.
* Source/JavaScriptCore/bytecode/AccessCase.h:
(JSC::SharedJITStubSet::Hash::Key::Key): Deleted.
(JSC::SharedJITStubSet::Hash::Key::isHashTableDeletedValue const): Deleted.
(JSC::SharedJITStubSet::Hash::hash): Deleted.
(JSC::SharedJITStubSet::Hash::equal): Deleted.
(JSC::SharedJITStubSet::Searcher::Translator::hash): Deleted.
(JSC::SharedJITStubSet::Searcher::Translator::equal): Deleted.
(JSC::SharedJITStubSet::PointerTranslator::hash): Deleted.
(JSC::SharedJITStubSet::PointerTranslator::equal): Deleted.
(JSC::SharedJITStubSet::add): Deleted.
(JSC::SharedJITStubSet::remove): Deleted.
(JSC::SharedJITStubSet::find): Deleted.
* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::setupWithUnlinkedBaselineCode):
(JSC::CodeBlock::forEachStructureStubInfo):
(JSC::CodeBlock::resetBaselineJITData):
* Source/JavaScriptCore/bytecode/InlineAccess.cpp:
(JSC::InlineAccess::resetStubAsJumpInAccess):
(JSC::InlineAccess::resetStubAsJumpInAccessNotUsingInlineAccess):
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::InlineCacheHandler::InlineCacheHandler):
(JSC::m_watchpoints):
(JSC::InlineCacheHandler::createSlowPath):
(JSC::InlineCacheCompiler::generateSlowPathHandler):
(JSC::InlineCacheCompiler::regenerate):
(JSC::PolymorphicAccess::visitWeak const):
(JSC::InlineCacheHandler::aboutToDie):
(JSC::InlineCacheHandler::visitWeak const):
(JSC::PolymorphicAccess::PolymorphicAccess): Deleted.
(JSC::PolymorphicAccess::~PolymorphicAccess): Deleted.
(JSC::PolymorphicAccess::aboutToDie): Deleted.
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.h:
(JSC::AccessGenerationResult::handler const):
(JSC::AccessGenerationResult::AccessGenerationResult):
(JSC::AccessGenerationResult::operator== const): Deleted.
(JSC::AccessGenerationResult::operator bool const): Deleted.
(JSC::PolymorphicAccess::containsPC const): Deleted.
* Source/JavaScriptCore/bytecode/StructureStubInfo.cpp:
(JSC::StructureStubInfo::aboutToDie):
(JSC::StructureStubInfo::addAccessCase):
(JSC::StructureStubInfo::visitWeakReferences):
(JSC::StructureStubInfo::containsPC const):
(JSC::StructureStubInfo::initializeFromUnlinkedStructureStubInfo):
(JSC::StructureStubInfo::initializeFromDFGUnlinkedStructureStubInfo):
(JSC::SharedJITStubSet::getMegamorphic const):
(JSC::SharedJITStubSet::setMegamorphic):
(JSC::SharedJITStubSet::getSlowPathHandler const):
(JSC::SharedJITStubSet::setSlowPathHandler):
* Source/JavaScriptCore/bytecode/StructureStubInfo.h:
(JSC::StructureStubInfo::offsetOfHandler):
(JSC::SharedJITStubSet::Hash::Key::Key):
(JSC::SharedJITStubSet::Hash::Key::isHashTableDeletedValue const):
(JSC::SharedJITStubSet::Hash::hash):
(JSC::SharedJITStubSet::Hash::equal):
(JSC::SharedJITStubSet::Searcher::Translator::hash):
(JSC::SharedJITStubSet::Searcher::Translator::equal):
(JSC::SharedJITStubSet::PointerTranslator::hash):
(JSC::SharedJITStubSet::PointerTranslator::equal):
(JSC::SharedJITStubSet::add):
(JSC::SharedJITStubSet::remove):
(JSC::SharedJITStubSet::find):
* Source/JavaScriptCore/dfg/DFGJITCode.cpp:
(JSC::DFG::JITData::JITData):
(JSC::DFG::JITData::tryInitialize):
* Source/JavaScriptCore/dfg/DFGJITCode.h:
(JSC::DFG::JITData::tryCreate):
* Source/JavaScriptCore/dfg/DFGJITCompiler.cpp:
(JSC::DFG::JITCompiler::loadConstant):
(JSC::DFG::JITCompiler::loadStructureStubInfo):
(JSC::DFG::JITCompiler::addStructureStubInfo):
* Source/JavaScriptCore/dfg/DFGJITCompiler.h:
* Source/JavaScriptCore/dfg/DFGSlowPathGenerator.h:
(JSC::DFG::slowPathICCall):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
(JSC::DFG::SpeculativeJIT::compileDeleteById):
(JSC::DFG::SpeculativeJIT::compileDeleteByVal):
(JSC::DFG::SpeculativeJIT::compileInById):
(JSC::DFG::SpeculativeJIT::compileInByVal):
(JSC::DFG::SpeculativeJIT::compileHasPrivate):
(JSC::DFG::SpeculativeJIT::compilePutByVal):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::cachedGetById):
(JSC::DFG::SpeculativeJIT::cachedGetByIdWithThis):
(JSC::DFG::SpeculativeJIT::compileGetByVal):
(JSC::DFG::SpeculativeJIT::compile):
(JSC::DFG::SpeculativeJIT::compileGetByValWithThis):
(JSC::DFG::SpeculativeJIT::compileEnumeratorPutByVal):
* Source/JavaScriptCore/jit/BaselineJITCode.cpp:
(JSC::BaselineJITData::BaselineJITData):
* Source/JavaScriptCore/jit/BaselineJITCode.h:
* Source/JavaScriptCore/jit/BaselineJITRegisters.h:
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::addUnlinkedStructureStubInfo):
* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JITCall.cpp:
(JSC::JIT::emit_op_iterator_open):
(JSC::JIT::emit_op_iterator_next):
* Source/JavaScriptCore/jit/JITCode.h:
(JSC::StructureStubInfoIndex::StructureStubInfoIndex):
* Source/JavaScriptCore/jit/JITInlineCacheGenerator.cpp:
(JSC::JITInlineCacheGenerator::generateDFGDataICFastPath):
(JSC::JITInlineCacheGenerator::generateBaselineDataICFastPath):
(JSC::generateGetByIdInlineAccessBaselineDataIC):
(JSC::generateGetByIdInlineAccessDFGDataIC):
(JSC::JITGetByIdGenerator::generateDFGDataICFastPath):
(JSC::JITGetByIdWithThisGenerator::generateDFGDataICFastPath):
(JSC::generatePutByIdInlineAccessBaselineDataIC):
(JSC::JITPutByIdGenerator::generateDFGDataICFastPath):
(JSC::generateInByIdInlineAccessBaselineDataIC):
(JSC::JITInByIdGenerator::generateDFGDataICFastPath):
(JSC::generateGetByIdInlineAccessDataIC): Deleted.
(JSC::generatePutByIdInlineAccessDataIC): Deleted.
(JSC::generateInByIdInlineAccessDataIC): Deleted.
* Source/JavaScriptCore/jit/JITInlineCacheGenerator.h:
* Source/JavaScriptCore/jit/JITInlines.h:
(JSC::JIT::loadConstant):
(JSC::JIT::loadStructureStubInfo):
* Source/JavaScriptCore/jit/JITOpcodes.cpp:
(JSC::JIT::emit_op_instanceof):
* Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
(JSC::JIT::emit_op_get_by_val):
(JSC::JIT::emit_op_get_private_name):
(JSC::JIT::emit_op_set_private_brand):
(JSC::JIT::emit_op_check_private_brand):
(JSC::JIT::emit_op_put_by_val):
(JSC::JIT::emit_op_put_private_name):
(JSC::JIT::emit_op_del_by_id):
(JSC::JIT::emit_op_del_by_val):
(JSC::JIT::emit_op_try_get_by_id):
(JSC::JIT::emit_op_get_by_id_direct):
(JSC::JIT::emit_op_get_by_id):
(JSC::JIT::emit_op_get_by_id_with_this):
(JSC::JIT::emit_op_put_by_id):
(JSC::JIT::emit_op_in_by_id):
(JSC::JIT::emit_op_in_by_val):
(JSC::JIT::emitHasPrivate):
(JSC::JIT::emit_op_get_by_val_with_this):
(JSC::JIT::emit_op_enumerator_get_by_val):
(JSC::JIT::emit_op_enumerator_put_by_val):
* Source/JavaScriptCore/runtime/VM.cpp:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/ButterflyArray.h: Added.
(WTF::ButterflyArray::ButterflyArray):
(WTF::ButterflyArray::createImpl):
(WTF::ButterflyArray::allocationSize):
(WTF::ButterflyArray::offsetOfLeadingSize):
(WTF::ButterflyArray::offsetOfTrailingSize):
(WTF::ButterflyArray::offsetOfTrailingData):
(WTF::ButterflyArray::memoryOffsetForDerived):
(WTF::ButterflyArray::leadingSpan):
(WTF::ButterflyArray::leadingSpan const):
(WTF::ButterflyArray::trailingSpan):
(WTF::ButterflyArray::trailingSpan const):
(WTF::ButterflyArray::operator delete):
(WTF::ButterflyArray::~ButterflyArray):
(WTF::ButterflyArray::leadingData):
(WTF::ButterflyArray::leadingData const):
(WTF::ButterflyArray::trailingData):
(WTF::ButterflyArray::trailingData const):
* Source/WTF/wtf/CMakeLists.txt:

Canonical link: <a href="https://commits.webkit.org/269685@main">https://commits.webkit.org/269685@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/3508a9db6fc39850321f41b7c02852f3d46085e8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23169 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1256 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24290 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25088 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21416 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2727 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23708 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22269 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23410 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/835 "Found 1 new test failure: http/tests/site-isolation/basic-iframe-render-output.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20084 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25932 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/661 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20970 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27135 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/20159 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21102 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21234 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25004 "Found 1 new API test failure: /WebKitGTK/TestWebViewEditor:/webkit/WebKitWebView/cut-copy-paste/editable (failure)") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/22532 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/668 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18493 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/29908 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/602 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6188 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1079 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/29860 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2962 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/877 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6081 "Passed tests") |

### CODE REVIEW

File: Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp [hunk 1]
```diff
@@ -1139,6 +1145,42 @@ MacroAssemblerCodeRef<JITThunkPtrTag> InlineCacheCompiler::generateSlowPathCode(
...
+    result->m_jumpTarget = CodePtr<NoPtrTag> { codeRef.retaggedCode<NoPtrTag>().dataLocation<uint8_t*>() + prologueSizeInBytesDataIC }.template retagged<JITStubRoutinePtrTag>();
+    return result;
+}
+
+Ref<InlineCacheHandler> InlineCacheCompiler::generateSlowPathHandler(VM& vm, AccessType accessType)
+{
+    ASSERT(!isCompilationThread());
+    ASSERT(Options::useHandlerIC());
+    if (auto handler = vm.m_sharedJITStubs->getSlowPathHandler(accessType))
+        return handler.releaseNonNull();
```
Comment (kmiller68) [inline Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:1178]: Why does this release and not increment?

File: Source/JavaScriptCore/jit/GPRInfo.h [hunk 2]
```diff
@@ -1173,7 +1187,7 @@ class PreferredArgumentImpl {
             GPRInfo::argumentGPR2, GPRInfo::argumentGPR3,
             GPRInfo::regT2,        GPRInfo::regT3,
             GPRInfo::regT4,        GPRInfo::regT5,
-            GPRInfo::regT6,        GPRInfo::regT7);
```
Comment (kmiller68) [inline Source/JavaScriptCore/jit/GPRInfo.h:1176]: What's up with this change?

File: Source/WTF/wtf/ButterflyArray.h [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+#include <wtf/Vector.h>
+
+namespace WTF {
+
+// ButterflyArray offers the feature trailing and leading array in the derived class.
+// We can allocate a memory like the following layout.
+//
+//     [ Leading Array ][  DerivedClass  ][ Trailing Array ]
+template<typename Derived, typename LeadingType, typename TrailingType>
+class ButterflyArray {
```
Comment (kmiller68) [inline Source/WTF/wtf/ButterflyArray.h:39]: Nice!

**Review by kmiller68 — APPROVED** [review]
r=me.

File: Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp (hunk 1 above)
Comment (Constellation) [reply Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:1178]: It is just converting RefPtr to Ref.

File: Source/JavaScriptCore/jit/GPRInfo.h (hunk 2 above)
Comment (Constellation) [reply Source/JavaScriptCore/jit/GPRInfo.h:1176]: They are not correctly working with the current use. Just fixed.

**Pull request merged**